api/desecapi/serializers.py

```diff
@@ -1,6 +1,71 @@
 from rest_framework import serializers
-from desecapi.models import Domain, Donation, User
+from desecapi.models import Domain, Donation, User, RRset
 from djoser import serializers as djoserSerializers
+import json
+
+
+class JSONSerializer(serializers.Field):
+    def to_representation(self, obj):
+        return json.loads(obj)
+
+    def to_internal_value(self, data):
+        return json.dumps(data)
+
+
+class RecordsSerializer(JSONSerializer):
+    def to_internal_value(self, records):
+        if not all(isinstance(record, str) for record in records):
+            msg = 'Incorrect type. Expected a list of strings'
+            raise serializers.ValidationError(msg)
+
+        # https://lists.isc.org/pipermail/bind-users/2008-April/070148.html
+        if not len(records) < 4092:
+            msg = 'Records too long. Must be less than 4092 characters, but was %d'
+            raise serializers.ValidationError(msg % len(records))
+
+        return super().to_internal_value(records)
+
+
+class FromContext(object):
+    def __init__(self, value_fn):
+        self.value_fn = value_fn
+
+    def set_context(self, serializer_field):
+        self.value = self.value_fn(serializer_field.context)
+
+    def __call__(self):
+        return self.value
+
+
+class RRsetSerializer(serializers.ModelSerializer):
+    # Disallow moving RRset to different zone
+    domain = serializers.SlugRelatedField(read_only=True, slug_field='name',
+                                          default=FromContext(
+                                              lambda context: Domain.objects.get(name=context['view'].kwargs['name'], owner=context['request'].user.pk)
+                                          ))
+
+    # Disallow renaming RRset (RRset.update_pdns() does not cope with this)
+    subname = serializers.CharField(allow_blank=True, read_only=True,
+                                 default=FromContext(
+                                     lambda context: context['request'].data.get('subname', '').lower()
+                                 ))
+
+    # Disallow changing RRset type (RRset.update_pdns() does not cope with this)
+    type = serializers.CharField(read_only=True,
```

Comment:
- does not shadow (referenced as `self.type`)
- moved write restriction to model layer, https://github.com/desec-io/desec-stack/commit/16ef52342c657c0b1dfdb11099ccb4c78b7c4eae